tests/transforms/test_decompose_transform.py

```diff
@@ -0,0 +1,240 @@
+
+
+class NoMatOp(Operation):
+    """Dummy operation for expanding circuit."""
+
+    # pylint: disable=arguments-renamed, invalid-overridden-method
+    @property
+    def has_matrix(self):
+        return False
+
+    def decomposition(self):
+        return [qml.PauliX(self.wires), qml.PauliY(self.wires)]
+
+
+class NoMatNoDecompOp(Operation):
+    """Dummy operation for checking check_validity throws error when
+    expected."""
+
+    # pylint: disable=arguments-renamed, invalid-overridden-method
+    @property
+    def has_matrix(self):
+        return False
+
+
+class InfiniteOp(Operation):
+    """An op with an infinite decomposition."""
+
+    num_wires = 1
+
+    def decomposition(self):
+        return [InfiniteOp(*self.parameters, self.wires)]
+
+
+class TestDecompose:
+    """Unit tests for decompose function"""
+
+    gate_set_inputs = [None, "RX", ["RX"], ("RX",), {"RX"}, qml.RX, [qml.RX], (qml.RX,), {qml.RX}]
+
+    iterables_test = [
+        (
+            [qml.Hadamard(0)],
+            {qml.RX, qml.RZ},
+            [qml.RZ(qnp.pi / 2, 0), qml.RX(qnp.pi / 2, 0), qml.RZ(qnp.pi / 2, 0)],
+            None,
+        ),
+        (
+            [qml.SWAP(wires=[0, 1])],
+            {qml.CNOT},
+            [qml.CNOT([0, 1]), qml.CNOT([1, 0]), qml.CNOT([0, 1])],
+            None,
+        ),
+        (
+            [qml.Toffoli([0, 1, 2])],
+            {qml.Toffoli},
+            [qml.Toffoli([0, 1, 2])],
+            None,
+        ),
+        (
+            [qml.measurements.MidMeasureMP(0)],
+            {},
+            [qml.measurements.MidMeasureMP(0)],
+            "MidMeasureMP",
+        ),
+    ]
+
+    callables_test = [
+        (
+            [qml.Hadamard(0)],
+            lambda op: "RX" in op.name,
+            [qml.RZ(qnp.pi / 2, 0), qml.RX(qnp.pi / 2, 0), qml.RZ(qnp.pi / 2, 0)],
+            "RZ",
+        ),
+        (
+            [qml.Toffoli([0, 1, 2])],
+            lambda op: len(op.wires) <= 2,
+            [
+                qml.Hadamard(wires=[2]),
+                qml.CNOT(wires=[1, 2]),
+                qml.ops.op_math.Adjoint(qml.T(wires=[2])),
+                qml.CNOT(wires=[0, 2]),
+                qml.T(wires=[2]),
+                qml.CNOT(wires=[1, 2]),
+                qml.ops.op_math.Adjoint(qml.T(wires=[2])),
+                qml.CNOT(wires=[0, 2]),
+                qml.T(wires=[1]),
+                qml.T(wires=[2]),
+                qml.CNOT(wires=[0, 1]),
+                qml.Hadamard(wires=[2]),
+                qml.T(wires=[0]),
+                qml.ops.op_math.Adjoint(qml.T(wires=[1])),
+                qml.CNOT(wires=[0, 1]),
+            ],
+            None,
+        ),
+    ]
+
+    @pytest.mark.parametrize("gate_set", gate_set_inputs)
+    def test_different_input_formats(self, gate_set):
+        """Tests that gate sets of different types are handled correctly"""
+        tape = qml.tape.QuantumScript([qml.RX(0, wires=[0])])
+        decompose(tape, gate_set=gate_set)
```

Comment:
I'm a little confused by this test. What are we checking for here? Should we check the output still has the RX in it?

But even if we had `gateset={}`, we'd still have an RX in the output, just with a warning about RX not having a decomposition.

Note that we currently don't treat warnings as errors.  While it would be wonderful to start treating warnings as errors, we currently raise so many warnings in tests that it would take too much time to fix.

Reply:
I intended for the test to check that users can "accidentally" supply the wrong `gate_set` input type and not have any problems. 

Since it's a more user-facing function, I added type converting guardrails to `decompose` to ensure that if the user accidentally specified just a string, `gate_set = "RX"`, the decomposition would still occur.

But in hindsight, [...], (...) and {...} are all sets so they should work by default. I was only really concerned if the user specified None, str or type.

Reply:
I've modified the test to make sure that the tape is still the same after the decomposition despite the different input types.